### PR TITLE
Simplify multi-monitor presentations

### PR DESCRIPTION
**Pull request by wentasah**
This is a followup to #2200. I addressed the comments there and also fixed problems I encountered with the previous version (look for Okular in the commit message). I also rebased the patches to next (hence the new PR). I run the testsuite, but it failed even without my commits.

### CODE REVIEW

**Comment by wentasah** [comment]
This also seems to fix the problem with freerdp mentioned in  #2116.

File: include/randr.h [hunk 1]
```diff
@@ -79,6 +79,13 @@ Output *get_output_by_name(const char *name);
  */
 Output *get_output_containing(unsigned int x, unsigned int y);
 
+/**
+ * Returns the active output which spans exactly the area specified by
+ * rect or NULL if there is no output like this.
+ *
+ */
+Output *get_output_by_rect(Rect rect);
```
Comment (stapelberg) [inline include/randr.h]: I think the name of this function could be a bit clearer. Currently, from reading just the name, the function might also return the first output which contains (parts of, or the entirety of) the given rect.

How about `get_output_with_dimensions(Rect rect)` instead?

File: src/con.c [hunk 2]
```diff
@@ -1158,6 +1158,21 @@ void con_move_to_workspace(Con *con, Con *workspace, bool fix_coordinates, bool
...
 
 /*
+ * Moves the given container to the currently focused container on the
+ * visible workspace on the given output.
+ *
+ */
+void con_move_to_output(Con *con, Output *output) {
+    /* fs is either the visible workspace or fullscreen window on the
+     * given output */
+    Con *fs = con_get_fullscreen_con(output_get_content(output->con), CF_OUTPUT);
```
Comment (stapelberg) [inline src/con.c]: The following pattern is more consistent with the i3 code base:

```c
Con *ws = NULL;
GREP_FIRST(ws, output_get_content(output->con), workspace_is_visible(child));
```

File: src/manage.c [hunk 3]
```diff
@@ -359,8 +359,16 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
     if (xcb_reply_contains_atom(state_reply, A__NET_WM_STATE_FULLSCREEN)) {
         /* If this window is already fullscreen (after restarting!), skip
          * toggling fullscreen, that would drop it out of fullscreen mode. */
-        if (fs != nc)
+        if (fs != nc) {
+            Output *output = get_output_by_rect((Rect){geom->x, geom->y, geom->width, geom->height});
+            /* If the requested window geometry spans the whole area
+             * of an output, move the window to that output. This is
+             * needed for multi-monitor presentations to work out of
```
Comment (stapelberg) [inline src/manage.c]: Can you include the name of the program (LibreOffice?) with which you tested this? “multi-monitor presentations” in general could be implemented using override_redirect, hence bypassing the window manager entirely. So, strictly speaking, it’s not necessary for the window manager to cooperate :).

File: src/handlers.c [hunk 4]
```diff
@@ -719,6 +719,21 @@ static void handle_client_message(xcb_client_message_event_t *event) {
...
+                     * output, move the window to that output. This is
+                     * needed for multi-monitor presentations to work
+                     * out of the box. */
+                    Output *output = get_output_by_rect(con->geometry);
+                    struct timeval second_ago;
+
+                    gettimeofday(&second_ago, NULL);
+                    second_ago.tv_sec--;
+
+                    if (output && _i3_timercmp(con->map_time, second_ago, >= ))
```
Comment (stapelberg) [inline src/handlers.c]: I dislike the fact that time is involved in this piece of code.

Can we remove this code entirely instead? Applications should map their window with the proper geometry and hints, I’d say.

File: include/randr.h (hunk 1 above)
Comment (Airblader) [inline include/randr.h]: :) We talked about the name before, this name was my quick suggestion. I do like yours even better.

File: src/handlers.c (hunk 4 above)
Comment (Airblader) [inline src/handlers.c]: Yes, this definitely can't be in here. If LibreOffice doesn't set the hints properly, it's a bug against LibreOffice. It sets the geometry to signal fullscreen, so if it wants fullscreen, it should set that hint before mapping.

File: src/manage.c [hunk 5]
```diff
@@ -359,8 +359,16 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
...
+             * the box. */
+            if (output)
```
Comment (Airblader) [inline src/manage.c]: @stapelberg I'm personally a fan of being explicit and writing `!= NULL`. I think this is currently not consistent in the codebase. What do you prefer? I'd like to get this more consistent.

**Comment by Airblader** [comment]
We'll also need tests for this.

File: src/handlers.c (hunk 4 above)
Comment (wentasah) [inline src/handlers.c]: LibreOffice (at least recent versions) does not need this. It sets the fullscreen hint before mapping its windows. I need this for pdfpc. I'll file a bug against it.

**Comment by wentasah** [comment]
Somewhere, I read that Xephyr does not support multiple outputs. How to
test this feature then? If there is some way, can you recommend me a
similar existing test from which I can start?

**Comment by Airblader** [comment]
i3 has "fake outputs" used for tests. Just search for "fake-outputs" in the testcases. The `new-test` script in `testcases/` also takes a `-m` flag for this.

**Comment by wentasah** [comment]
This is an updated version addressing the above comments. It includes a new test and is rebased onto current "next".

File: src/con.c [hunk 6]
```diff
@@ -1158,6 +1158,22 @@ void con_move_to_workspace(Con *con, Con *workspace, bool fix_coordinates, bool
...
+ *
+ */
+void con_move_to_output(Con *con, Output *output) {
+    Con *ws = NULL;
+    GREP_FIRST(ws, output_get_content(output->con), workspace_is_visible(child));
+    if (ws != NULL) {
+        DLOG("Moving con %p to output %s\n", con, output->name);
+        con_move_to_workspace(con, ws, false, false, false);
+    } else {
+        DLOG("No visible workspace on output %s\n", output->name);
```
Comment (Airblader) [inline src/con.c]: I think we should use `assert(ws != NULL)` here, or at least log this as an error. We should *always* have a visible workspace on an active output.

Comment (wentasah) [inline src/con.c]: I also think that the else part cannot happen, but in all other places in the code, the else branch was present. Anyway, I'm pushing a new version with assert.

**Comment by wentasah** [comment]
Ping. I there anything I can do to move this patch further?

**Comment by wentasah** [comment]
Ping. I there anything I can do to move this patch further?

**Comment by stapelberg** [comment]
Sorry, this fell through the cracks. Merged now, thanks!

**Pull request merged**